doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,257 @@
+# Disable user indexes mode
+
+## Summary
+
+When users have views causing OOM crashes, we want to provide a mechanism that
+still lets the cloud team boot into Materialize and take actions to remediate
+the issue, e.g. drop the OOMing view. To accomplish this, we'll introduce a new
+boot mode that starts Materialize as normal with the exception of not shipping
+any dataflows for user-created objects, such as materialized views.
+
+Ideally, we'll also provide a mechanism that lets users enable dataflows for
+their objects while Materialize is running using a statement such as `ALTER
+INDEX <index name> SET ENABLED`.
+
+### Using "Disable user indexes mode"
+
+This feature's initial iteration will find most use from  expert users of
+Materialize, e.g. the cloud team. Using it, they'll be able to boot Materialize
+in this mode, and use the system's reporting metrics to determine which indexes
+caused the user grief, e.g. an OOM crash, and drop it. They can then turn on the
+remaining indexes one-by-one to ensure the system doesn't crash. If it doesn't,
+they can then reboot with all of the indexes running.
+
+The challenging thing here is using the reporting metrics to determine which
+indexes caused the crashes. We lack comprehensive education around these
+metrics, so it's unclear how most users could suss out where there problems are.
+
+One clever solution to this is to introduce a `--disable-recent-indexes N`
+feature, which I think would make for a great user experience as a follow-on to
+the main proposal in this doc. There are more details about this in the
+**Alternatives** section.
+
+## Goals
+
+- Support booting without starting any user-created dataflows.
+- Introduce a `--disable-user-indexes` flag to Materialize that lets users
+  easily turn on this mode.
+- Add support for `ALTER INDEX ... SET ENABLED` to enable dataflows for objects
+  while Materialilze is running, and lift the restrictions imposed on
+  catalog-only mode for objects that use those dataflows.
+
+  Note that this statement will be "intrinsically idempotent," so `ALTER INDEX
+  ... SET ENABLED` will always succeed, even if Materialize was not booted in
+  catalog only mode or its dataflows are already active. This matches PG's
+  behavior for `ALTER` statements like `ALTER TABLE a ALTER COLUMN x SET NOT
+  NULL`, which you can run repeatedly.
+- Add no more than a single boolean check to the code path of any operation
+  occurring outside of catalog-only mode.
+
```

Comment:
What's the motivation behind this last goal? If the motivation is performance, you can definitely scratch this goal. Parsing, planning, and optimizing a SQL query is going to absolutely dwarf any possible implementation of catalog-only mode.